robocop/checkers/misc.py

```diff
@@ -1566,3 +1644,47 @@ def check_for_complex_condition(self, condition_token, node_name, left_side, var
                 col=position,
                 end_col=position + len(variable) + len(right_side),
             )
+
+
+class TestAndKeywordOrderChecker(VisitorChecker):
+    reports = ("test-case-section-out-of-order", "keyword-section-out-of-order")
+
+    def __init__(self):
+        self.rules_by_node_type = {}
+        self.expected_order = {}
+        super().__init__()
+
+    def visit_File(self, node):  # noqa
+        self.rules_by_node_type = {Keyword: "keyword-section-out-of-order", TestCase: "test-case-section-out-of-order"}
+        self.expected_order = {
+            Keyword: self.param("keyword-section-out-of-order", "sections_order"),
+            TestCase: self.param("test-case-section-out-of-order", "sections_order"),
+        }
+        self.generic_visit(node)
+
+    def check_order(self, node):
+        if type(node) not in self.rules_by_node_type:
+            raise ValueError("Invalid node type")
+        already_used_sections = []
+        for subnode in node.body:
+            try:
+                subnode_type = subnode.type
+            except AttributeError:
+                continue
+            if subnode_type not in self.expected_order[type(node)]:
+                continue
+            this_node_expected_order = self.expected_order[type(node)].index(subnode_type)
+            for item in already_used_sections:
+                if this_node_expected_order < self.expected_order[type(node)].index(item):
+                    self.report(
+                        self.rules_by_node_type[type(node)],
+                        section_name=subnode_type,
+                        recommended_order=", ".join(self.expected_order[type(node)]),
+                        node=subnode,
+                        col=subnode.col_offset + 1,
+                        end_col=subnode.end_col_offset + 1,
+                    )
+                    break
+            already_used_sections.append(subnode_type)
```

Comment:
Since we're appending to already_used_sections for each element in node.body, all iterations (for example for keyword that has 5 items) would look like:

1 item:
  no iterations since already_used_sections = 0
 2nd item:
   1 iterations
  3nd item
    2 iterations 
    
    And so on - and we are checking the order of items that were already checked. It could be good idea to extract used_sections in one for loop, and then check the order:
    
 for subnode in node.body:
    # get type if no exception and in expected_order
    # save it to other list
 for item in other list:
     check order here

Reply:
You are right, this was far away from optimal.
I decided to rework this in a bit different way: without list of already used sub-nodes, but only to remember maximal order indicator of already used sub-nodes. Thanks to that even second loop is not needed.

Reply:
There is one downside of using one flag for it - if for example one setting is out of order, it may report invalid order for whole body (ie 20 lines) instead of reporting that single setting which is out of order. It could be workarounded by reporting this rule only once per test/keyword (but if several settings are out of order, it will hide it)

Reply:
Results returned by current and previous implementation are the same. Every test's / keyword's line is reported if is after one of previous lines when when should be before. `break` in second loop in previous implementation was only to not report that given line is after multiple lines.
To reduce number of reported issues we could:
- report this rule only once per test/keyword as you suggested.
- find issues when iterating through node.body and remember all issues, then do similar but by iterating reversed node.body), then compare number of issues found in each direction and call self.report() for shorter set of issues.
or we can leave it.
Please decide